web/src/typeahead_helper.ts

```diff
@@ -796,10 +802,117 @@ export let sort_group_setting_options = ({
     return setting_options.slice(0, MAX_ITEMS);
 };
 
+export let sort_group_setting_options = ({
+    users,
+    query,
+    groups,
+    target_group,
+}: {
+    users: UserPillData[];
+    query: string;
+    groups: UserGroupPillData[];
+    target_group: UserGroup | undefined;
+}): (UserPillData | UserGroupPillData)[] =>
+    sort_setting_options({
+        users,
+        query,
+        groups,
+        compare_setting_options: compare_group_setting_options,
+        target_group,
+    });
+
 export function rewire_sort_group_setting_options(value: typeof sort_group_setting_options): void {
     sort_group_setting_options = value;
 }
 
+export function compare_stream_setting_options(
+    option_a: UserPillData | UserGroupPillData,
+    option_b: UserPillData | UserGroupPillData,
+): number {
+    if (option_a.type === "user_group") {
+        const user_group_a = user_groups.get_user_group_from_id(option_a.id);
+        if (user_group_a.name === "role:members") {
+            return -1;
+        }
+    }
+    if (option_b.type === "user_group") {
+        const user_group_b = user_groups.get_user_group_from_id(option_b.id);
+        if (user_group_b.name === "role:members") {
+            return 1;
+        }
+    }
+
+    if (option_a.type === "user_group" && option_b.type === "user") {
+        const user_group_a = user_groups.get_user_group_from_id(option_a.id);
+
+        if (user_group_a.is_system_group) {
+            return 1;
+        }
+        return -1;
+    }
+
+    if (option_b.type === "user_group" && option_a.type === "user") {
+        const user_group_b = user_groups.get_user_group_from_id(option_b.id);
+        if (user_group_b.is_system_group) {
+            return -1;
+        }
+        return 1;
+    }
+
+    if (option_a.type === "user_group" && option_b.type === "user_group") {
+        const user_group_a = user_groups.get_user_group_from_id(option_a.id);
+        const user_group_b = user_groups.get_user_group_from_id(option_b.id);
+
+        if (user_group_a.is_system_group && !user_group_b.is_system_group) {
+            return 1;
+        }
+
+        if (user_group_b.is_system_group && !user_group_a.is_system_group) {
+            return -1;
+        }
+
+        if (user_group_a.name < user_group_b.name) {
```

Comment:
`util.strcmp` should be used here.

Reply:
Tests fail if I change this. I think I'll merge what you have here, and you can do a follow-up PR changing to use this -- while you're at it, can you also audit the other places in this file where we compare strings with `<`, which generally produces weird results?